Add option for updating Notepad++ at exit

xomx opened the pull request:
And add `/relaunchNppAfterSilentInstall` into installer (ref: https://github.com/notepad-plus-plus/notepad-plus-plus/pull/15280#issuecomment-2246816010)

Fix #16601 , Fix #13749 , Fix #10317 , Fix #8495 , Fix #8457 , Fix #3755

![npp-Pref-MISC-AutoUpdate-AtExit](https://github.com/user-attachments/assets/0c60629f-7c3b-49e1-81c3-8b1cc531296e)

xomx commented:
@ehawman-rosenberg
and whoever wants to participate on testing this PR

Now this PR is in a testing phase only, in the code is hardcoded some N++ version value for GUP (to be able to always bring up GUP update dialog), also the GUP itself will need a slight modification if this PR will be accepted:

![npp-GUP-UpdateAvailable-msgbox-at-exit](https://github.com/user-attachments/assets/00605696-2837-494d-a72e-b3c5a80472bc)

For testing you will need to download this PR GitHub Action artifact, here for the x64 Release Notepad++.exe binary:
https://github.com/notepad-plus-plus/notepad-plus-plus/actions/runs/15413907360/artifacts/3248546905
and replace your current (rename the original one to e.g. notepad++.exe.ori) N++ main executable `C:\Program Files\Notepad++\notepad++.exe`.

Then also open your `%APPDATA%\Notepad++\config.xml` file (backup it 1st to e.g. `config.xml.ori`) and check/modify the `noUpdate` item there:

`        <GUIConfig name="noUpdate" intervalDays="15" nextUpdateDate="20250617" atAppExit="no">yes</GUIConfig>`

The `nextUpdateDate` has to be lower than your current comp date, e.g. `"20250601"` (06 01 ... June 1st), otherwise there will not be the GUP updater for testing the feature.

Try it with and without the new `At exit` option and let me know ok/bad.

xomx commented:
@ehawman-rosenberg 
> One small hiccup is that saying "No" with atAppExist="no" updates nextUpdateDate, while saying "No" with atAppExit="yes" does not.

Good catch! Fixed.

> Functionally it's pretty much exactly what I was hoping for.

Ok, so I will update the PR heading post to invite more potential users to test & comment.

> adding a little checkbox or button to the up-front GUP dialog that says "Update on Close" would be really nice, but of course we don't want it to look like a soundboard. Maybe replace "Never" with opening the Preferences to the appropriate location?

While it is possible to do so, it would encompass adding a new custom N++ WM message for the inter-process communication (GUP->N++). IMO not worth it. And with the `At Exit` option ON, it's even useless (in the same way as pressing the `Never` GUP-dlg btn now, as I've already explained above - a N++ wnd capable of receiving such msgs is not behind at the N++ exit time...).

> ... Preferences UI rearranging img ...

This is of course possible but with this I will wait for the possible project maintainer opinion, I'm trying to keep this PR as simple as possible.

> ... Preferences UI new `Update silently` checkbox possibility img ...

Yes, I thought about such option. But it requires thorough simultaneous debugging of both N++ & GUP from my side to be sure there wouldn't be any problems. Maybe later, when I have more time.

xomx commented:
@donho 
> Chrome - and if you take the same method to download & delay the installation

Of course not, I read your previous POV on this in the relevant FR issues.

> (have not yet seen your code), then it's a big no.

Then you will be surprised how simple my sln is.

tl;dr summary - I took the existing updating code before entering the N++ main msg-loop

https://github.com/notepad-plus-plus/notepad-plus-plus/blob/8435308b028e93dc99a9e88ddba089a544e6dc50/PowerEditor/src/winmain.cpp#L651-L705

and without any optimization, I also copied it after leaving the msg-loop (with some minor N++ exit-specific changes).
The rest of the PR is for one more Preferences checkbox and its corresponding config.xml stuff.

donho commented:
@xomx 
> What is my gist from all the users' FR posts in the linked issues - "I'd like to have the N++ updates ON, just do not bother me at all at the N++ start, where I usually have something urgent to do 1st - that is why I just launched the app!"

OK, I think it's legit.
But where is your gist then? When I go:
https://gist.github.com/xomx
I see: ***xomx doesn’t have any public gists yet. ***

>  I could try to add the complete "updates-silentization" to this PR too.

Hmm... very interesting. But how can you workaround the system popup for the application installation?

donho commented:
@ehawman-rosenberg 
> So an alternative of, say, showing an "Update Available" icon (✨) in the top bar next to ? (for manual post-launch updates) wouldn't work, correct?

I don't really like this solution. So we will focus on current PR with the current update interface.

xomx commented:
@donho 
> But where is your gist then?

Sorry I was not clear, @ehawman-rosenberg is right, these were just my own words trying to summarize how I feel when reading the user FR posts on this matter.

> > I could try to add the complete "updates-silentization" to this PR too.
> 
> Hmm... very interesting. But how can you workaround the system popup for the application installation?

Just did a quick test.
- I added one more (now hardcoded) param at the N++ wWinMain updater places (in this PR before & after the main msg-loop):  `updaterParams += L" -unattended";`
- added the same param parsing to the GUP project wWinMain and if detected, it simply skip the YesYessilentNoNever-dlg and directly use the `FLAG_NSIS_SILENT_INSTALL_PARAM` instead

Then I only see (at N++ start or end - depending on the `AtExit` checkbox):
- the NSIS installer downloading progress-wnd (can also be hidden)
- NSIS installer UAC-prompt (meanwhile it cannot be suppressed but I plan to also look at this later, see the https://github.com/notepad-plus-plus/notepad-plus-plus/issues/15182#issuecomment-2916190015

So - until our NSIS installer supports per-user installations and/or until the user installs N++ to a protected location (such as the standard %PROGRAMFILES%), the system UAC popup will always be there.

donho commented:
@xomx 
> So - until our NSIS installer supports per-user installations and/or until the user installs N++ to a protected location (such as the standard %PROGRAMFILES%), the system UAC popup will always be there.

Yes, I'm talking about UAC popup - once we use installer for update usage, the totally silent update will never come true.

xomx commented:
@donho 
> once we use installer for update usage, the totally silent update will never come true.

Yes, that's true. Unless we e.g. do not include an installation of some N++ system update helper service in the N++ NSIS installer.

I think that this PR should remain simple as is and cover only the originally requested installation on close. If I add the above mentioned `-unattended` option to suppress the YesYessilentNoNever-dlg, it can create a user confusion as "why suddenly there is that UAC-prompt at the N++ exit?!" (this is also an argument against the complete silentization - when there will be e.g. an regression/bug in the new N++ version silently installed, users could start to be bewildered as why suddenly their N++ doesn't work...)

xomx commented:
@donho 
> By helper service you mean a daemon-like background process right?

Yes.

> Or do you mean such service has already Admin privilege?

Yes, this. A standard Windows system service, a list of which can be viewed by e.g. running the "services.msc". Here is the short summary for such a way (I'm not saying we're going down this path, now just for the info):
- we write a tiny Windows system service module (easy, it is enough to write a small Windows console app with only the winmain() and the [LPSERVICE_MAIN_FUNCTION](https://learn.microsoft.com/en-us/windows/win32/api/winsvc/nc-winsvc-lpservice_main_functionw) callback func, nothing more)
- install this service from our already elevated NSIS installer by: `sc.exe create <npp-service-name> binPath= "<FullPathTo\npp-service.exe>"`
- now the main part, we move the current notepad++.exe wWinMain update-checking code to that new npp-service.exe module
- and that's it, because now the code for checking for N++ updates (and possible launching of the NSIS-installer) will be run with SYSTEM privileges, so no UAC popup will appear...



> Yes, I feel the same way.

Ok. So let's proceed to finalizing this PR.
Do you like the current Preferences>MISC UI or should I change it or rearrange somehow?
This is the current layout, there is IMO still enough space for the translations:

![npp-Pref-MISC-AtExit-layout](https://github.com/user-attachments/assets/b1ca0de0-29da-425a-a1f8-39a8fc7018ce)

Do you like the "At exit" or should it rather be "On close"?

donho commented:
> Yes, this. A standard Windows system service, a list of which can be viewed by e.g. running the "services.msc". Here is the short summary for such a way (I'm not saying we're going down this path, now just for the info):

I see. But personally I don't like this idea, not only do I try to keep the simplicty of application - but I believe also users don't want to see more (useless?) processes stay in the memory. As a result, it's unlikely to be implemented. 

> Do you like the "At exit" or should it rather be "On close"?

I think it would be better to change as "system tray" and "rendering mode", but with the label and combox reversed:

![image](https://github.com/user-attachments/assets/dcf88427-af6e-41a4-a545-fd95fbfd9ba2)

and the combobox contains 3 values:

```
Auto-updater:    Disable               
                 Enable on Notepad++ startup
                 Enable on Notepad++ exit
```
`_nppGUI._autoUpdateOpt._doAutoUpdate` won't be boolean anymore, but rather a integer (or enum if you prefer) which will contains 0 (disable), 1 (Enable on Notepad++ startup), 2 (Enable on Notepad++ exit).

The retro-compatibility should be kept, so we keep read "noUpdate" on startup for the convertion, and read & write new entry "autoUpdater" with value 0, 1 or 2.

Do you think it's OK, or you have other suggestions?

xomx commented:
@donho 

> the combobox contains 3 values:

Ok, auto-update UI switched to combobox

> `_nppGUI._autoUpdateOpt._doAutoUpdate` won't be boolean anymore, but rather a integer (or enum if you prefer) which will contains 0 (disable), 1 (Enable on Notepad++ startup), 2 (Enable on Notepad++ exit).

Done.

> The retro-compatibility should be kept, so we keep read "noUpdate" on startup for the convertion, and read & write new entry "autoUpdater" with value 0, 1 or 2.

Done. Code tested with the older configs and ok.

---

I've already removed from the PR the temporary hardcoded fake version value for the GUP, so now to test - there has to be a real newer N++ version available on the net.

If you think that all is ok, I will add the translation strings.

Also I am going to do a small fix for the GUP project (it will disable the "Never" button in the YesYessilentNoNever-dlg in case that there is no N++ wnd behind for receiving that command msg).

donho commented:
@xomx 
Just tried the new feature of the PR, works as expected, but only one thing:
While user chooses "Enable on Notepad++ exit", what they want, IMO, is after clicking on "Yes Silently" and few popup, there's no Notepad++ relaunched. It's not the case. Though it's as expected for dev, I am wondering if it's as expected for users. 
If the most users don't expect relaunch of Notepad++ after updating on exit, we should modify the NSIS script in this PR, and modify WinGUp as well for being included in the next version.
What do you think?

xomx commented:
@donho 

> While users choose "Enable on Notepad++ exit", what they want, IMO, is after clicking on "Yes Silently" and few popup, there's no Notepad++ relaunched.

Very good assumption, I agree. If someone intentionally closes an app, they most likely want it to stay closed.

> What do you think?

I remembered this request a while ago: https://github.com/notepad-plus-plus/notepad-plus-plus/pull/15280#issuecomment-2246816010

> "... I would like to automatically re-open npp only if the installation closed it, but if it was already closed it should not open it."

So what if we only modify the N++ installer NSI-scripts to always remember if there was a simultaneously running notepad++.exe instance at the installer time? The N++ mutex detection part is already there anyway. Then use simple logic - restart N++ at installer exit only when such a previously running instance was detected. With the new N++ `autoupdate-on-close` selected, there will not be any N++ instance running at the installer time, so also no N++ relaunch...

donho commented:
@xomx 
> So what if we only modify the N++ installer NSI-scripts to always remember if there was a simultaneously running notepad++.exe instance at the installer time? The N++ mutex detection part is already there anyway. Then use simple logic - restart N++ at installer exit only when such a previously running instance was detected. With the new N++ autoupdate-on-close selected, there will not be any N++ instance running at the installer time, so also no N++ relaunch...

Sorry, I don't really follow you - could you elaborate ?

xomx commented:
@donho Ok. I am on a short mountain trip. Will be back on this in 2-3 days.

xomx commented on PowerEditor/installer/nppSetup.nsi:
```diff
@@ -385,6 +386,7 @@ Section -FinishSection
   Call writeInstallInfoInRegistry
   IfSilent 0 theEnd
   	${If} $runNppAfterSilentInstall == "true"
+	${AndIf} $runningNppDetected == "true" ; relaunch only if Notepad++ was already running before the installation
```
@donho 
> Do you confirm above description?

Almost - but in the "on exit scenario" instead of the
>  then WinGUp DOESN'T detect the instance of Notepad++, and launches Installer with "/runNppAfterSilentInstall" parameter.

it should be

> then after the WinGUp launches Installer with `FLAG_NSIS_SILENT_INSTALL_PARAM` (== `/closeRunningNpp /S /runNppAfterSilentInstall`), the Installer DOESN'T detect the instance of Notepad++ running and thus does not relaunch the Notepad++ app after the updating installation.

---

More details for "on startup scenario":
- when Notepad++ launches WinGUp on startup (before entering the N++ main msg-loop), the main N++ app-wnd does not exist yet (it will be created immediately in the following `notepad_plus_plus.init`)
- when YesSilent of WinGUp button is clicked, the WinGUp uses the `FLAG_NSIS_SILENT_INSTALL_PARAM` (== `/closeRunningNpp /S /runNppAfterSilentInstall`)
- so, for the current NSIS-installer, this could be a problem because for the N++ app running detection (`$runningNppDetected == "true"`) it uses also the `FindWindow` WINAPI (in the  `FindAndCloseOrTerminateRunningNpp` macro for the param `/closeRunningNpp`)
- but recently, the WinGUp has been already fixed by me to wait 5 secs (now in a new current PR modified to 3 secs only) for a detection of the N++ mutex to be present, so the N++ app has enough time to finish its init (creating the necessary N++ main-wnd), before the NSIS-installer uses its `FindWindow` WINAPI for searching for the running N++ instances...


It's getting pretty complicated, isn't it? :-)
But the main thing is that there should not be any race condition.

donho replied:
@xomx 
Sorry for the late reply, it takes me a little time to read the code and remember how it was implemented.
It becomes really quite complexe with the time (evolution).

The silent auto-update procedure for the goal of this PR, could be simplified into the following step:
* Notepad++ launches WinGUp
* WinGUp launches nppInstaller
* Installer launches (**or not** - if "at exit" option is enabled) Notepad++

So far the code is almost ready for me:
1. in Notepad++, you have added the code for the on exit installation.
2. in Installer,(script NSIS), we have already `${If} $runNppAfterSilentInstall == "true" Call LaunchNpp`

So it's evident for me, that instead of add the confusion logic into installer, we should only to distiguish no "on exit" & with "on exit", to pass "/runNppAfterSilentInstall" or not to the installer.

Do you have any solution with logic I mention above?

donho replied:
After checking WinGUp code, it might be a solution:
```cpp
				if (!g_hAppWnd)
				{
					::EnableWindow(::GetDlgItem(hWndDlg, IDCANCEL), FALSE); // no app-wnd to sent the updates disabling message signal
					FLAG_NSIS_SILENT_INSTALL_PARAM[] = L"/S";
				}
```

pryrt replied:
The "only if" phrase at the end does make it slightly confusing.  I propose the following:
```
/relaunchNppAfterSilentInstall: If Notepad++ was running when silent install was initiated, it will automatically run the newly-installed Notepad++ after installation is complete (New to v8.8.2.)
    Only works if /S is also specified.
```

donho replied:
@xomx 
OK, please make both **runNppAfterSilentInstall** & **relaunchNppAfterSilentInstall** so installer will be more flexible for Notepad++ deployment.

xomx commented:
@donho 
> * Add new parameter "relaunchNppAfterSilentInstall" for the installer.
> 
> * Add the translation

All done.

Thanks for handling the WinGUp update part.

donho reviewed (changes requested):
Please consider the suggestions.

xomx commented on PowerEditor/src/winmain.cpp:
```diff
@@ -755,14 +757,90 @@ int WINAPI wWinMain(_In_ HINSTANCE hInstance, _In_opt_ HINSTANCE /*hPrevInstance
+				nppGui._autoUpdateOpt._intervalDays = 0 - nppGui._autoUpdateOpt._intervalDays;
+			nppGui._autoUpdateOpt._nextUpdateDate = Date(nppGui._autoUpdateOpt._intervalDays);
+			nppParameters.createXmlTreeFromGUIParams();
+			nppParameters.saveConfig_xml();
```
@donho 
I think we cannot, there is the updating of the local var `doUpdateNpp`.
Or do you mean not a void but a "bool launchUpdater()" func or similar?
IMO not worth it. 

> It's shame that we have to duplicate the code...

I know but I tried to do not touch much the updater code logic. The overall updating logic is complicated enough.

I'm in favor of going back to this sometime and changing it more fundamentally (and halving the length of the code used e.g. via that new `launchUpdater(but-with-some-params)`).

donho replied:
@xomx
IMO, the following code can be restored:
 
```cpp
    if (doUpdateNpp) // check more detail
    {
        Date today(0);

        if (today < nppGui._autoUpdateOpt._nextUpdateDate)
            doUpdateNpp = false;
    }

	if (doUpdatePluginList)
	{
		// TODO: detect update frequency
		// Due to the code signing problem, the Plugin List cannot be updated independently of Notepad++ for now.
	}
```

So refactoring of "launchUpdater()" can be done.

donho replied:
Then make the following function:
```cpp
void launchUpdater(const std::wstring& updaterFullPath, const std::wstring& updaterDir)
{
	std::wstring updaterParams = L"-v";
	updaterParams += VERSION_INTERNAL_VALUE;

	NppParameters& nppParameters = NppParameters::getInstance();

	if (nppParameters.archType() == IMAGE_FILE_MACHINE_AMD64)
	{
		updaterParams += L" -px64";
	}
	else if (nppParameters.archType() == IMAGE_FILE_MACHINE_ARM64)
	{
		updaterParams += L" -parm64";
	}

	Process updater(updaterFullPath.c_str(), updaterParams.c_str(), updaterDir.c_str());
	updater.run();

	// Update next update date
	NppGUI& nppGui = nppParameters.getNppGUI();
	if (nppGui._autoUpdateOpt._intervalDays < 0) // Make sure interval days value is positive
		nppGui._autoUpdateOpt._intervalDays = 0 - nppGui._autoUpdateOpt._intervalDays;
	nppGui._autoUpdateOpt._nextUpdateDate = Date(nppGui._autoUpdateOpt._intervalDays);
}
```

donho replied:
```cpp
	SecurityGuard securityGuard;
	bool isSignatureOK = securityGuard.checkModule(updaterFullPath, nm_gup);

	if (TheFirstOne && isUpExist && isGtXP && isSignatureOK && doUpdateNpp && !updateAtExit)
	{
		launchUpdater(updaterFullPath, updaterDir);
	}


	...


	doUpdateNpp = nppGui._autoUpdateOpt._doAutoUpdate != NppGUI::autoupdate_disabled; // refresh, maybe user activated these opts in Preferences
	updateAtExit = nppGui._autoUpdateOpt._doAutoUpdate == NppGUI::autoupdate_on_exit; // refresh
	if (!isException && !nppParameters.isEndSessionCritical() && TheFirstOne && isUpExist && isGtXP && isSignatureOK && doUpdateNpp && updateAtExit)
	{
		launchUpdater(updaterFullPath, updaterDir);
	}

	return static_cast<int>(msg.wParam);
}

```

donho reviewed (changes requested):
Please reconsider the suggestion.

donho reviewed (changes requested):
Please consider the following suggestion.

donho commented on PowerEditor/src/winmain.cpp:
```diff
@@ -755,14 +725,28 @@ int WINAPI wWinMain(_In_ HINSTANCE hInstance, _In_opt_ HINSTANCE /*hPrevInstance
+		{
+			// for updating the nextUpdateDate in the already saved config.xml
+			nppParameters.createXmlTreeFromGUIParams();
+			nppParameters.saveConfig_xml();
```
> I moved also the Date today(0); ... check into the launchUpdater(...) and change that new func to return bool for detecting the need to update the config.xml at the on-exit case.

You are right to do so. When I wrote "IMO, the following code can be restored", I was tiered and I've mistaken with the following code:
```cpp
	// Update next update date
	NppGUI& nppGui = nppParameters.getNppGUI();
	if (nppGui._autoUpdateOpt._intervalDays < 0) // Make sure interval days value is positive
		nppGui._autoUpdateOpt._intervalDays = 0 - nppGui._autoUpdateOpt._intervalDays;
	nppGui._autoUpdateOpt._nextUpdateDate = Date(nppGui._autoUpdateOpt._intervalDays);
```
The above code was to me to be moved out, in order to prevent rewriting config.xml (again, I was tired, with a little bit alcohol/beer I admit :) ).

It turns out that rewriting config.xml is inevitable. I hope the nightmare of corrupted configuration files won't come back to haunt us.

xomx replied:
Yes, when I saw the requested changes I was like WTF, did I overlooked something?

I just had a looong and tiring day and at 4am in the morning I wanted to finally get some sleep. Instead of sleeping, I was finishing this PR because I won't have time for it a while now. That's why I wrote you had to review, because of I was already zombified when I finished it.

donho replied:
> Yes, when I saw the requested changes I was like WTF, did I overlooked something?

😅

> I just had a looong and tiring day and at 4am in the morning I wanted to finally get some sleep. Instead of sleeping, I was finishing this PR because I won't have time for it a while now. That's why I wrote you had to review, because of I was already zombified when I finished it.

My apology for the bad quality code view of last night!
Thank you for the insistence for the right thing.
